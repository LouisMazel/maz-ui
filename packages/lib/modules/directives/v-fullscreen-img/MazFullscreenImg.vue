<template>
  <div
    ref="FullscreenImgElement"
    role="button"
    class="m-fullscreen-img"
    tabindex="0"
    @click.stop="close"
    @keypress.esc.prevent="close"
  >
    <MazSpinner v-show="showLoader" class="m-fullscreen-img-loader maz-h-16" />

    <button
      v-if="loadedOnce && hasMultipleInstances"
      type="button"
      class="m-fullscreen-btn --next"
      @click.stop="nextPreviousImage('next')"
    >
      <ChevronLeft class="maz-rotate-180 maz-text-3xl" />
    </button>
    <button
      v-if="loadedOnce && hasMultipleInstances"
      type="button"
      class="m-fullscreen-btn --previous"
      @click.stop="nextPreviousImage('previous')"
    >
      <ChevronLeft class="maz-text-3xl" />
    </button>

    <button type="button" class="m-fullscreen-btn --close" @click="close">
      <XMark class="maz-text-3xl" />
    </button>

    <div class="m-fullscreen-img-scroller">
      <!-- eslint-disable-next-line vuejs-accessibility/no-static-element-interactions, vuejs-accessibility/click-events-have-key-events -->
      <img
        ref="ImgElement"
        :src="currentSrc"
        :alt="currentAlt ?? undefined"
        tabindex="0"
        :class="[imageZoomClasses]"
        @load="onImageLoaded"
        @click.stop="zoom ? toggleZoom() : undefined"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
  import { computed, defineAsyncComponent, onBeforeUnmount, ref, onMounted, watch } from 'vue'

  const emits = defineEmits(['close', 'previous', 'next', 'before-close'])

  const MazSpinner = defineAsyncComponent(() => import('./../../../components/MazSpinner.vue'))
  const XMark = defineAsyncComponent(() => import('./../../../icons/x-mark.svg'))
  const ChevronLeft = defineAsyncComponent(() => import('./../../../icons/chevron-left.svg'))

  export type Props = {
    src: string
    clickedElementBounds?: {
      top: number
      left: number
      height: number
      width: number
    }
    offset?: number
    animation?: {
      duration?: number
      easing?: string
    }
    openInstanceClass?: string
    clickedElement: HTMLElement
    destroy?: () => void
    alt?: string
    zoom?: boolean
  }

  const props = withDefaults(defineProps<Props>(), {
    zoom: true,
    offset: undefined,
    destroy: undefined,
    alt: undefined,
    animation: () => ({
      duration: 300,
      easing: 'ease-in-out',
    }),
    clickedElementBounds: undefined,
    openInstanceClass: 'm-fullscreen-img-instance',
  })

  const imageLoaded = ref(false)
  const showLoader = ref(false)
  const loadedOnce = ref(false)
  const hasMultipleInstances = ref(false)
  const isZoomed = ref(false)

  const currentElement = ref(props.clickedElement)
  const currentElementBounds = computed(() => props.clickedElement.getBoundingClientRect())
  const isLandscapeImage = computed(
    () => currentElementBounds.value.height < currentElementBounds.value.width,
  )

  const currentSrc = ref(props.src)
  const currentAlt = ref<string | null | undefined>(props.alt)

  const FullscreenImgElement = ref<HTMLDivElement>()
  const ImgElement = ref<HTMLImageElement>()

  const imageZoomClasses = computed(() => {
    if (isZoomed.value) {
      return `--is-zoomed maz-cursor-zoom-out`
    }

    return `maz-cursor-zoom-in`
  })

  function onImageLoaded() {
    imageLoaded.value = true
    showLoader.value = false
    loadedOnce.value = true
  }

  watch(
    () => loadedOnce.value,
    (value) => {
      if (value) {
        openFullscreen()
      }
    },
  )

  function close() {
    emits('before-close')
    closeFullscreen()
  }

  function keydownLister(e: KeyboardEvent) {
    if (e.key === 'Escape') {
      e.preventDefault()
      close()
    }

    if (e.key === 'ArrowLeft' || e.key === 'ArrowRight') {
      e.preventDefault()
      nextPreviousImage(e.key === 'ArrowRight' ? 'next' : 'previous')
    }
  }

  function addClassToDocument() {
    document.documentElement.classList.add('--m-fullscreen-open')
  }

  function removeClassFromDocument() {
    document.documentElement.classList.remove('--m-fullscreen-open')
  }

  function getAllInstances(): HTMLElement[] {
    return [...document.querySelectorAll('.m-fullscreen-img-instance')] as HTMLElement[]
  }

  function getNewInstanceIndex(allInstances: HTMLElement[], newInstanceIndex: number): number {
    return newInstanceIndex < 0
      ? allInstances.length - 1
      : newInstanceIndex >= allInstances.length
        ? 0
        : newInstanceIndex
  }

  async function useNextInstance(currentInstance: HTMLElement, nextInstance: HTMLElement) {
    currentInstance.classList.remove(props.openInstanceClass)
    nextInstance.classList.add(props.openInstanceClass)

    const src: string | null = nextInstance.getAttribute('data-src')
    const alt: string | null = nextInstance.getAttribute('data-alt')

    currentAlt.value = alt
    if (src) {
      currentSrc.value = src
    }

    imageLoaded.value = false

    if (!imageLoaded.value) {
      showLoader.value = true
    }
  }

  function nextPreviousImage(which: 'next' | 'previous'): void {
    const currentInstance: HTMLElement | null = document.querySelector(
      `.m-fullscreen-img-instance.${props.openInstanceClass}`,
    )

    if (currentInstance) {
      const allInstances = getAllInstances()
      const currentInstanceIndex = allInstances.indexOf(currentInstance)
      const newInstanceIndex =
        which === 'next' ? currentInstanceIndex + 1 : currentInstanceIndex - 1

      const nextInstance = allInstances[getNewInstanceIndex(allInstances, newInstanceIndex)]
      currentElement.value = nextInstance

      if (nextInstance) {
        useNextInstance(currentInstance, nextInstance)
      }

      emits(which)

      if (isZoomed.value) {
        setZoomStyles(ImgElement.value as HTMLElement)
      } else {
        setEndAnimationStyles()
      }
    }
  }

  function setZoomStyles(imgElement: HTMLElement) {
    if (isLandscapeImage.value) {
      imgElement.style.width = `${window.innerWidth}px`
      imgElement.style.removeProperty('height')
    } else {
      imgElement.style.height = `${window.innerHeight}px`
      imgElement.style.removeProperty('width')
    }

    imgElement.style.removeProperty('top')
    imgElement.style.removeProperty('left')
  }

  async function toggleZoom() {
    const imgElement = ImgElement.value as HTMLImageElement

    if (isZoomed.value) {
      isZoomed.value = !isZoomed.value
      setEndAnimationStyles()
    } else {
      isZoomed.value = !isZoomed.value
      setZoomStyles(imgElement)
    }
  }

  function runAnimation(frames: Keyframe[] | PropertyIndexedKeyframes) {
    return ImgElement.value?.animate(frames, {
      duration: props.animation.duration, // Durée de l'animation en millisecondes
      easing: props.animation.easing, // Fonction d'interpolation pour l'animation inverse
    })
  }

  function getPositions(offset = props.offset ?? 0) {
    const originalWidth =
      currentElement.value instanceof HTMLImageElement
        ? currentElement.value.naturalWidth
        : currentElement.value.clientWidth
    const originalHeight =
      currentElement.value instanceof HTMLImageElement
        ? currentElement.value.naturalHeight
        : currentElement.value.clientHeight

    // Obtenez les dimensions de la fenêtre
    const windowWidth = window.innerWidth
    const windowHeight = window.innerHeight

    // Calculez le facteur d'échelle pour conserver les proportions
    const scale = Math.min(
      (windowWidth - 2 * offset) / originalWidth,
      (windowHeight - 2 * offset) / originalHeight,
    )

    // Calculez les coordonnées pour centrer l'image
    const centerX = (windowWidth - originalWidth * scale) / 2
    const centerY = (windowHeight - originalHeight * scale) / 2

    return {
      originalWidth,
      originalHeight,
      scale,
      centerX,
      centerY,
    }
  }

  function getAnimationFrames({ trigger }: { trigger: 'open' | 'close' }) {
    const { originalWidth, originalHeight, scale, centerX, centerY } = getPositions()

    const { top, left, width, height } = currentElementBounds.value

    const frames = [
      {
        top: `${top}px`,
        left: `${left}px`,
        width: `${width}px`,
        height: `${height}px`,
        opacity: 0,
      },
      {
        top: `${centerY}px`,
        left: `${centerX}px`,
        width: `${originalWidth * scale}px`,
        height: `${originalHeight * scale}px`,
        opacity: 1,
      },
    ]

    return {
      frames: trigger === 'open' ? frames : frames.reverse(),
    }
  }

  function setEndAnimationStyles() {
    const { centerX, centerY, originalHeight, originalWidth, scale } = getPositions()

    const finalStyles = {
      top: `${centerY}px`,
      left: `${centerX}px`,
      width: `${originalWidth * scale}px`,
      height: `${originalHeight * scale}px`,
    }

    if (!ImgElement.value) {
      throw console.error('[maz-ui](vFullscreenImg) ImgElement is not defined')
    }

    Object.assign(ImgElement.value.style, finalStyles)
  }

  function openFullscreen() {
    const { frames } = getAnimationFrames({
      trigger: 'open',
    })

    // Utilisez la fonction animate pour gérer l'animation avec les nouvelles coordonnées

    const openAnimation = runAnimation(frames)

    // Ajoutez un gestionnaire d'événements pour détecter la fin de l'animation
    if (!openAnimation) {
      throw console.error('[maz-ui](vFullscreenImg) open animation is not defined')
    }

    openAnimation.onfinish = () => {
      setEndAnimationStyles()
    }
  }

  function closeFullscreen() {
    const { frames } = getAnimationFrames({
      trigger: 'close',
    })

    const closeAnimation = runAnimation(frames)

    if (!closeAnimation) {
      throw console.error('[maz-ui](vFullscreenImg) close animation is not defined')
    }

    // Ajoutez un gestionnaire d'événements pour détecter la fin de l'animation inverse
    closeAnimation.onfinish = () => {
      // Retirez l'élément du DOM
      emits('close')
      FullscreenImgElement.value?.remove()
      props.destroy?.()
    }
  }

  function onResizeWindow() {
    if (!isZoomed.value) {
      setEndAnimationStyles()
    }
  }

  onMounted(() => {
    showLoader.value = true
    document.addEventListener('keydown', keydownLister)
    window.addEventListener('resize', onResizeWindow)
    addClassToDocument()
    hasMultipleInstances.value = getAllInstances().length > 1
  })

  onBeforeUnmount(() => {
    document.removeEventListener('keydown', keydownLister)
    window.removeEventListener('resize', onResizeWindow)
    removeClassFromDocument()
  })
</script>

<style lang="postcss">
  html.--m-fullscreen-open {
    @apply maz-h-screen maz-overflow-hidden;
  }
</style>

<style lang="postcss" scoped>
  .m-fullscreen-img {
    @apply maz-fixed maz-inset-0 maz-z-default-backdrop maz-h-screen maz-w-screen maz-items-center maz-bg-overlay maz-outline-none maz-backdrop-blur;

    .m-fullscreen-img-scroller {
      @apply maz-flex maz-h-screen maz-w-screen maz-overflow-auto maz-flex-center;
    }

    img {
      @apply maz-fixed maz-z-[1051] maz-select-none maz-object-contain maz-object-center maz-outline-none;
    }

    .m-fullscreen-img-loader {
      @apply maz-absolute maz-left-1/2 maz-top-1/2 maz-z-10 maz--translate-x-1/2 maz--translate-y-1/2 maz-transform maz-text-6xl;
    }

    .m-fullscreen-btn {
      @apply maz-absolute maz-z-10 maz-flex maz-h-20 maz-w-[7%] maz-min-w-[5em] maz-cursor-pointer maz-p-4 maz-transition-colors maz-duration-200;

      &.--close {
        @apply maz-right-0 maz-top-0 maz-items-start maz-justify-end;
      }

      &.--previous {
        @apply maz-left-0 maz-top-1/2 maz-h-screen maz--translate-y-1/2 maz-transform maz-items-center maz-justify-start;
      }

      &.--next {
        @apply maz-right-0 maz-top-1/2 maz-h-screen maz--translate-y-1/2 maz-transform maz-items-center maz-justify-end;
      }
    }
  }
</style>
