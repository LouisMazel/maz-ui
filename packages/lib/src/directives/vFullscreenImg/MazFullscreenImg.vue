<script lang="ts" setup>
import type { HTMLAttributes } from 'vue'
import { MazChevronLeft, MazXMark } from '@maz-ui/icons'
import { checkAvailability } from '@maz-ui/utils/src/utils/checkAvailability.js'
import { computed, onBeforeUnmount, onMounted, reactive, ref, watch } from 'vue'

import MazSpinner from '../../components/MazSpinner.vue'

const props = withDefaults(defineProps<MazFullscreenImgProps>(), {
  zoom: true,
  offset: undefined,
  destroy: undefined,
  alt: undefined,
  animation: () => ({
    duration: 300,
    easing: 'ease-in-out',
  }),
  clickedElementBounds: undefined,
  openInstanceClass: 'm-fullscreen-img-instance',
})

const emits = defineEmits(['close', 'previous', 'next', 'before-close'])

export interface MazFullscreenImgProps {
  src: string
  clickedElementBounds?: {
    top: number
    left: number
    height: number
    width: number
  }
  offset?: number
  animation?: {
    duration?: number
    easing?: string
  }
  openInstanceClass?: string
  clickedElement: HTMLElement
  destroy?: () => void
  alt?: string | null
  zoom?: boolean
}

const imageLoaded = ref(false)
const showLoader = ref(false)
const loadedOnce = ref(false)
const hasMultipleInstances = ref(false)
const isZoomed = ref(false)
const animationState = reactive({
  running: false,
  ended: false,
})

const currentClickedElement = ref(props.clickedElement)
const currentClickedElementBounds = computed(() => props.clickedElement.getBoundingClientRect())
const isLandscapeImage = ref()

const currentSrc = ref(props.src)
const currentAlt = ref<string | null | undefined>(props.alt)

const FullscreenImgElement = ref<HTMLDivElement>()
const ImgElement = ref<HTMLImageElement>()
const hideImage = ref(true)

const imageZoomClasses = computed<HTMLAttributes['class']>(() => {
  return {
    '--is-zoomed': isZoomed.value,
    '--invisible': hideImage.value,
    '--absolute': !isZoomed.value,
  }
})

function onImageLoaded() {
  if (ImgElement.value) {
    isLandscapeImage.value = ImgElement.value?.naturalWidth > ImgElement.value?.naturalHeight
  }

  imageLoaded.value = true
  showLoader.value = false
  loadedOnce.value = true
}

watch(
  loadedOnce,
  (value) => {
    if (value) {
      openFullscreen()
    }
  },
  { immediate: true },
)

function close() {
  emits('before-close')
  closeFullscreen()
}

function keydownLister(e: KeyboardEvent) {
  if (e.key === 'Escape') {
    e.preventDefault()
    close()
  }

  if (e.key === 'ArrowLeft' || e.key === 'ArrowRight') {
    e.preventDefault()
    nextPreviousImage(e.key === 'ArrowRight' ? 'next' : 'previous')
  }
}

function addClassToDocument() {
  document.documentElement.classList.add('--m-fullscreen-open')
}

function removeClassFromDocument() {
  document.documentElement.classList.remove('--m-fullscreen-open')
}

function getAllInstances(): HTMLElement[] {
  return [...document.querySelectorAll('.m-fullscreen-img-instance')] as HTMLElement[]
}

function getNewInstanceIndex(allInstances: HTMLElement[], newInstanceIndex: number): number {
  if (newInstanceIndex < 0) {
    return allInstances.length - 1
  }
  if (newInstanceIndex >= allInstances.length) {
    return 0
  }
  return newInstanceIndex
}

function useNextInstance(currentInstance: HTMLElement, nextInstance: HTMLElement) {
  currentInstance.classList.remove(props.openInstanceClass)
  nextInstance.classList.add(props.openInstanceClass)

  const src: string | null = nextInstance.getAttribute('data-src')
  const alt: string | null = nextInstance.getAttribute('data-alt')

  currentAlt.value = alt
  currentSrc.value = src ?? currentSrc.value
}

function nextPreviousImage(which: 'next' | 'previous'): void {
  hideImage.value = true

  const currentInstance: HTMLElement | null = document.querySelector(
    `.m-fullscreen-img-instance.${props.openInstanceClass}`,
  )

  if (currentInstance) {
    const allInstances = getAllInstances()
    const currentInstanceIndex = allInstances.indexOf(currentInstance)
    const newInstanceIndex
        = which === 'next' ? currentInstanceIndex + 1 : currentInstanceIndex - 1

    const nextInstance = allInstances[getNewInstanceIndex(allInstances, newInstanceIndex)]
    currentClickedElement.value = nextInstance

    if (nextInstance) {
      useNextInstance(currentInstance, nextInstance)
    }

    emits(which)

    imageLoaded.value = false
    showLoader.value = true

    checkAvailability(() => imageLoaded.value === true, () => {
      hideImage.value = false
      if (isZoomed.value) {
        setZoomStyles()
      }
      else {
        setEndAnimationStyles()
      }
    }, {
      expectedValue: true,
      interval: 100,
      maxAttempts: 50,
    })
  }
}

function setZoomStyles() {
  const imgElement = ImgElement.value

  if (!imgElement) {
    console.error('[maz-ui](vFullscreenImg) ImgElement is not defined')
    return
  }

  imgElement.style.removeProperty('max-width')
  imgElement.style.removeProperty('max-height')
  imgElement?.style.removeProperty('top')
  imgElement?.style.removeProperty('left')

  if (!isLandscapeImage.value) {
    imgElement.style.width = `100vw`
    imgElement.style.removeProperty('height')
  }
  else {
    imgElement.style.height = `100vh`
    imgElement.style.removeProperty('width')
  }
}

function toggleZoom() {
  if (isZoomed.value) {
    isZoomed.value = !isZoomed.value
    setEndAnimationStyles()
  }
  else {
    isZoomed.value = !isZoomed.value
    setZoomStyles()
  }
}

function runAnimation(frames: Keyframe[] | PropertyIndexedKeyframes) {
  animationState.running = true
  hideImage.value = false

  const animation = ImgElement.value?.animate(frames, {
    duration: props.animation.duration,
    easing: props.animation.easing,
  })

  if (!animation) {
    console.error('[maz-ui](vFullscreenImg) animation is not defined')
    animationState.running = false
    animationState.ended = true
    return
  }

  return animation
}

function getPositionsOfClikedElement(offset = props.offset ?? 0) {
  const width = currentClickedElement.value.clientWidth || 1
  const height = currentClickedElement.value.clientHeight || 1

  const windowWidth = window.innerWidth
  const windowHeight = window.innerHeight

  const scale = Math.min(
    (windowWidth - 2 * offset) / width,
    (windowHeight - 2 * offset) / height,
  )

  const centerX = (windowWidth - width * scale) / 2
  const centerY = (windowHeight - height * scale) / 2

  return {
    centerX,
    centerY,
    width,
    height,
    scale,
  }
}

function getAnimationFrames({ trigger }: { trigger: 'open' | 'close' }) {
  const { width, height, scale, centerX, centerY } = getPositionsOfClikedElement()

  const { top, left, width: clickedElementWidth, height: clickedElementHeight } = currentClickedElementBounds.value

  const frames = [
    {
      top: `${top}px`,
      left: `${left}px`,
      width: `${clickedElementWidth}px`,
      height: `${clickedElementHeight}px`,
      opacity: 0,
    },
    {
      top: `${centerY}px`,
      left: `${centerX}px`,
      width: `${width * scale}px`,
      height: `${height * scale}px`,
      opacity: 1,
    },
  ]

  return {
    frames: trigger === 'open' ? frames : frames.reverse(),
  }
}

function setEndAnimationStyles() {
  const { height, width, scale } = getPositionsOfClikedElement()

  const finalStyles = isLandscapeImage.value
    ? {
        width: `${width * scale}px`,
        maxHeight: `${height * scale}px`,
      }
    : {
        height: `${height * scale}px`,
        maxWidth: `${width * scale}px`,
      }

  if (!ImgElement.value) {
    console.error('[maz-ui](vFullscreenImg) ImgElement is not defined')
    return
  }

  if (isLandscapeImage.value) {
    ImgElement.value.style.removeProperty('height')
    ImgElement.value.style.removeProperty('maxHeight')
  }
  else {
    ImgElement.value.style.removeProperty('width')
    ImgElement.value.style.removeProperty('maxWidth')
  }

  Object.assign(ImgElement.value.style, finalStyles)
}

function openFullscreen() {
  const { frames } = getAnimationFrames({
    trigger: 'open',
  })

  const openAnimation = runAnimation(frames)

  if (!openAnimation) {
    console.error('[maz-ui](vFullscreenImg) open animation is not defined')
    setEndAnimationStyles()
    return
  }

  openAnimation.onfinish = () => {
    setEndAnimationStyles()
    animationState.running = false
    animationState.ended = true
  }
}

function closeFullscreen() {
  const { frames } = getAnimationFrames({
    trigger: 'close',
  })

  const closeAnimation = runAnimation(frames)

  function onFinish() {
    emits('close')
    FullscreenImgElement.value?.remove()
    props.destroy?.()
    animationState.running = false
    animationState.ended = true
  }

  if (!closeAnimation) {
    console.error('[maz-ui](vFullscreenImg) close animation is not defined')
    onFinish()
    return
  }

  closeAnimation.onfinish = onFinish
}

function onResizeWindow() {
  if (!isZoomed.value) {
    setEndAnimationStyles()
  }
}

onMounted(() => {
  showLoader.value = true
  document.addEventListener('keydown', keydownLister)
  window.addEventListener('resize', onResizeWindow)
  addClassToDocument()
  hasMultipleInstances.value = getAllInstances().length > 1
})

onBeforeUnmount(() => {
  document.removeEventListener('keydown', keydownLister)
  window.removeEventListener('resize', onResizeWindow)
  removeClassFromDocument()
})
</script>

<template>
  <div
    ref="FullscreenImgElement"
    role="button"
    class="m-fullscreen-img m-reset-css"
    tabindex="0"
    @click.stop="close"
    @keypress.esc.prevent="close"
  >
    <button
      v-if="loadedOnce && hasMultipleInstances"
      type="button"
      class="m-fullscreen-btn --next"
      @click.stop="nextPreviousImage('next')"
    >
      <MazChevronLeft class="maz-rotate-180" />
    </button>
    <button
      v-if="loadedOnce && hasMultipleInstances"
      type="button"
      class="m-fullscreen-btn --previous"
      @click.stop="nextPreviousImage('previous')"
    >
      <MazChevronLeft />
    </button>

    <button type="button" class="m-fullscreen-btn --close" @click="close">
      <MazXMark />
    </button>

    <div class="m-fullscreen-img-scroller">
      <!-- eslint-disable-next-line vuejs-accessibility/no-static-element-interactions, vuejs-accessibility/click-events-have-key-events -->
      <img
        ref="ImgElement"
        :src="currentSrc"
        :alt="currentAlt ?? undefined"
        tabindex="0"
        :class="[imageZoomClasses]"
        @load="onImageLoaded"
        @click.stop="zoom && toggleZoom()"
      >

      <MazSpinner v-show="showLoader" class="m-fullscreen-img-loader" />
    </div>
  </div>
</template>

<style lang="postcss">
  html.--m-fullscreen-open {
  @apply maz-h-screen maz-overflow-hidden;
}
</style>

<style lang="postcss" scoped>
  .m-fullscreen-img {
  @apply maz-fixed maz-inset-0 maz-z-default-backdrop maz-h-screen maz-w-screen maz-items-center maz-bg-overlay/5 maz-outline-none maz-backdrop-blur;

  .m-fullscreen-img-scroller {
    @apply maz-flex maz-h-screen maz-w-screen maz-overflow-auto maz-flex-center;
  }

  img {
    @apply maz-z-2 maz-outline-none maz-cursor-zoom-in maz-object-center maz-object-contain;

    &.--is-zoomed {
      @apply maz-cursor-zoom-out;
    }

    &.--invisible {
      @apply maz-invisible;
    }

    &.--absolute {
      @apply maz-absolute;
    }
  }

  .m-fullscreen-img-loader {
    @apply maz-absolute maz-text-2xl maz-z-15;
  }

  .m-fullscreen-btn {
    @apply maz-absolute maz-z-15 maz-flex maz-h-20 maz-w-[7%] maz-min-w-[5em] maz-cursor-pointer maz-p-4 maz-transition-colors maz-duration-200;

    svg {
      @apply maz-text-3xl maz-transition-transform maz-duration-300 maz-ease-in-out;
    }

    &:hover {
      svg {
        @apply maz-scale-150;
      }
    }

    &.--close {
      @apply maz-right-0 maz-top-0 maz-items-start maz-justify-end;
    }

    &.--previous {
      @apply maz-left-0 maz-top-1/2 maz-h-screen maz--translate-y-1/2 maz-transform maz-items-center maz-justify-start;
    }

    &.--next {
      @apply maz-right-0 maz-top-1/2 maz-h-screen maz--translate-y-1/2 maz-transform maz-items-center maz-justify-end;
    }
  }
}
</style>
