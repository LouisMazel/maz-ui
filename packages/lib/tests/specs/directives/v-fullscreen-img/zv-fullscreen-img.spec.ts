import MazFullscreenImg from '@modules/directives/v-fullscreen-img/MazFullscreenImg.vue'
import { shallowMount } from '@vue/test-utils'

const defaultProperties = {
  clickedElementBounds: {
    top: 10,
    left: 20,
    width: 30,
    height: 40,
  },
  clickedElement: document.createElement('img'),
  animated: true,
  scaleAnimation: false,
}

describe('MazFullscreenImg', () => {
  test('should render correctly when opened', async () => {
    const wrapper = shallowMount(MazFullscreenImg, {
      props: {
        src: 'https://via.placeholder.com/150',
        alt: 'placeholder image',
        openInstanceClass: 'zv-open',
        destroy: () => {},
        ...defaultProperties,
      },
    })

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(wrapper.find('.m-fullscreen-img-loader').isVisible()).toBe(true)
    expect(wrapper.find('.m-fullscreen-btn').exists()).toBe(true)
  })

  test('should render correctly with an alternative text', async () => {
    const alt = 'alternative text'

    const wrapper = shallowMount(MazFullscreenImg, {
      props: {
        src: 'https://via.placeholder.com/150',
        alt,
        openInstanceClass: 'zv-open',
        destroy: () => {},
        ...defaultProperties,
      },
    })

    // expect(wrapper.find('.m-fullscreen-img-wrapper').isVisible()).toBe(true)
    expect(wrapper.find('img').attributes('alt')).toBe(alt)
  })

  test('should close the component and destroy it correctly', async () => {
    const destroy = vi.fn()

    const wrapper = shallowMount(MazFullscreenImg, {
      props: {
        src: 'https://via.placeholder.com/150',
        alt: 'placeholder image',
        openInstanceClass: 'zv-open',
        destroy,
        ...defaultProperties,
      },
    })

    await wrapper.find('.m-fullscreen-btn.--close').trigger('click')

    expect(wrapper.emitted('before-close')).toBeTruthy()
  })
})
