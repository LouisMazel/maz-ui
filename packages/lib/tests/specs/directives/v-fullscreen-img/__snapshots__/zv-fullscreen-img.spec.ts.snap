// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazFullscreenImg > Should renders correctly with default props 1`] = `
"<div data-v-717706d4=\\"\\" role=\\"button\\" class=\\"m-fullscreen-img\\" tabindex=\\"0\\">
  <maz-spinner-stub data-v-717706d4=\\"\\" class=\\"m-fullscreen-img-loader maz-h-16\\" style=\\"display: none;\\"></maz-spinner-stub>
  <!--v-if-->
  <!--v-if--><button data-v-717706d4=\\"\\" type=\\"button\\" class=\\"m-fullscreen-btn --close\\">
    <x-mark-stub data-v-717706d4=\\"\\" class=\\"maz-text-2xl\\"></x-mark-stub>
  </button>
  <div data-v-717706d4=\\"\\" class=\\"m-fullscreen-img-scroller\\">
    <transition-stub data-v-717706d4=\\"\\" appear=\\"false\\" persisted=\\"true\\" css=\\"true\\"><img data-v-717706d4=\\"\\" id=\\"ZvImgElement\\" src=\\"https://via.placeholder.com/150\\" alt=\\"placeholder image\\" tabindex=\\"0\\" class=\\"zoomable cursor-zoom-out maz-h-screen maz-w-auto\\" style=\\"display: none;\\"></transition-stub>
  </div>
</div>"
`;
